File: src/main/java/scripting/AbstractPlayerInteraction.java
package scripting;

import client.Character;
import client.Client;
import client.effects.field.TrembleFieldEffect;
import client.effects.user.QuestEffect;
import client.interaction.storage.ItemStorageInteraction;
import client.inventory.ItemVariationType;
import client.inventory.item.templates.ItemEquipTemplate;
import client.inventory.item.templates.ItemTemplate;
import client.inventory.slots.ItemSlotEquip;
import client.messages.IncEXPMessage;
import client.messages.broadcast.types.AlertMessage;
import client.messages.broadcast.types.NoticeWithoutPrefixMessage;
import client.messages.quest.PerformQuestRecordMessage;
import client.player.quest.Quest;
import field.object.life.FieldMob;
import field.object.life.FieldMobTemplate;
import lombok.Getter;
import lombok.NonNull;
import lombok.RequiredArgsConstructor;
import managers.ItemManager;
import managers.MobManager;
import net.maple.packets.CharacterPackets;
import net.maple.packets.GuildPackets;
import org.graalvm.collections.Pair;
import scripting.npc.NPCScriptManager;

import java.util.AbstractMap;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.concurrent.TimeUnit;
import java.util.function.Function;

@RequiredArgsConstructor
@Getter
@SuppressWarnings("unused")
public abstract class AbstractPlayerInteraction {

    protected @NonNull final Client c;

    public Character getPlayer() {
        return c.getCharacter();
    }

    public void gainItem(int id, int quantity) {
        gainItemInternal(id, quantity);
        c.write(CharacterPackets.localEffect(new QuestEffect(id, quantity)));
    }

    public void massGainItem(int[][] items) {
        List<Pair<Integer, Integer>> entries = new ArrayList<>();
        Arrays.stream(items).forEach(item -> {
            gainItemInternal(item[0], item[1]);
            entries.add(Pair.create(item[0], item[1]));
        });
        c.write(CharacterPackets.localEffect(new QuestEffect(entries)));
    }

    private void gainItemInternal(int id, int quantity) {
        if (quantity > 0) {
            ItemTemplate item = ItemManager.getItem(id);
            if (item != null) {
                CharacterPackets.modifyInventory(getPlayer(),
                        i -> i.add(item, (short) quantity),
                        false);
            }
        } else {
            CharacterPackets.modifyInventory(getPlayer(),
                    i -> i.take(id, (short) -quantity),
                    false);
        }
    }

    public void gainMeso(int meso) {
        getPlayer().gainMeso(meso, true);
    }

    public boolean haveItem(int id, int quantity) {
        return getPlayer().getItemQuantity(id) >= quantity;
    }

    public void gainExp(int exp) {
        getPlayer().gainExp(exp);

        IncEXPMessage msg = new IncEXPMessage();
        msg.setExp(exp);
        msg.setOnQuest(true);
        getPlayer().write(CharacterPackets.message(msg));
    }

    public void warp(int id, String portal) {
        getPlayer().changeField(id, portal);
    }

    public void warp(int id) {
        getPlayer().changeField(id);
    }

    public int getMapId() {
        return getPlayer().getFieldId();
    }

    public void sendBlue(String message) {
        getPlayer().write(CharacterPackets.message(new NoticeWithoutPrefixMessage(message)));
    }

    public void setQuestProgress(int qid, int mob, String progress) {
        Quest quest = getPlayer().getQuests().get(qid);
        if (quest != null) {
            quest.progress(mob, Integer.parseInt(progress));
        }
    }

    public Quest getQuest(int id) {
        return getPlayer().getQuests().get(id);
    }

    public void openNpc(int id) {
        if (!NPCScriptManager.getInstance().converse(c, id)) {
            sendBlue("Npc does not have a script");
        }
    }

    public boolean isEquip(int id) {
        ItemTemplate template = ItemManager.getItem(id);
        if (template == null) return false;
        return template instanceof ItemEquipTemplate;
    }

    public ItemSlotEquip getEquipById(int id) {
        ItemTemplate template = ItemManager.getItem(id);
        if (template == null) return null;
        return (ItemSlotEquip) template.toItemSlot();
    }

    @SuppressWarnings({"unchecked"})
    public void gainStatItem(int id, Object obj) {
        AbstractMap<String, Integer> stats = (AbstractMap) obj;
        ItemSlotEquip equip = getEquipById(id);
        if (equip != null) {
            equip.setSTR(stats.get("STR").shortValue());
            equip.setDEX(stats.get("DEX").shortValue());
            equip.setLUK(stats.get("LUK").shortValue());
            equip.setINT(stats.get("INT").shortValue());
            equip.setPAD(stats.get("PAD").shortValue());
            equip.setMAD(stats.get("MAD").shortValue());
            equip.setACC(stats.get("ACC").shortValue());
            equip.setEVA(stats.get("EVA").shortValue());
            equip.setJump(stats.get("JUMP").shortValue());
            equip.setSpeed(stats.get("SPEED").shortValue());
            equip.setPDD(stats.get("PDD").shortValue());
            equip.setMDD(stats.get("MDD").shortValue());
            equip.setMaxHP(stats.get("HP").shortValue());
            equip.setMaxMP(stats.get("MP").shortValue());
            equip.setRUC(stats.get("SLOTS").byteValue());

            CharacterPackets.modifyInventory(getPlayer(), i -> i.add(equip), false);
        }
    }

    public void tremble(boolean heavy, int delay) {
        getPlayer().getField().broadcast(CharacterPackets.fieldEffect(new TrembleFieldEffect(heavy, delay)));
    }

    public void openNpcIn(int npc, int time, boolean dispose) {
        c.getCh().eventLoop().schedule(() -> {
            if (dispose) {
                c.setLastNpcClick(0);
                NPCScriptManager.getInstance().dispose(c);
            }
            NPCScriptManager.getInstance().converse(c, npc);
        }, time, TimeUnit.MILLISECONDS);
    }

    public void openNpcIn(int npc, int time) {
        openNpcIn(npc, time, true);
    }

    public void executeAfter(Function<AbstractPlayerInteraction, Void> func, int after) {
        c.getCh().eventLoop().schedule(() -> func.apply(this), after, TimeUnit.MILLISECONDS);
    }

    public void alert(String msg) {
        c.write(CharacterPackets.message(new AlertMessage(msg)));
    }

    public void openStorage(int npcId) {
        new ItemStorageInteraction(npcId, c.getStorage()).open(getPlayer());
    }

    public void changeGuildName(String name) {
        GuildPackets.changeGuildName(getPlayer(), name);
    }

    public void loadGuild() {
        if (getPlayer().getGuild() == null) return;
        c.write(GuildPackets.getLoadGuildPacket(getPlayer().getGuild()));
    }

    public FieldMobTemplate getMobTemplate(int id) {
        return MobManager.getMob(id);
    }
}

File: src/main/java/scripting/AbstractPlayerInteraction.kt
package scripting

import client.Character
import client.Client
import client.effects.field.TrembleFieldEffect
import client.effects.user.QuestEffect
import client.interaction.storage.ItemStorageInteraction
import client.inventory.item.slots.ItemSlotEquip
import client.inventory.item.templates.ItemEquipTemplate
import client.messages.IncEXPMessage
import client.messages.broadcast.types.AlertMessage
import client.messages.broadcast.types.NoticeWithoutPrefixMessage
import client.messages.broadcast.types.UtilDlgExMessage
import client.player.quest.Quest
import field.Field
import field.obj.drop.AbstractFieldDrop
import field.obj.life.FieldMob
import field.obj.life.FieldMobTemplate
import field.obj.life.FieldNPC
import managers.ItemManager
import managers.MobManager
import net.maple.packets.CharacterPackets.localEffect
import net.maple.packets.CharacterPackets.message
import net.maple.packets.CharacterPackets.modifyInventory
import net.maple.packets.FieldPackets.fieldEffect
import net.maple.packets.GuildPackets
import net.maple.packets.GuildPackets.getLoadGuildPacket
import net.server.Server
import org.graalvm.collections.Pair
import scripting.npc.NPCScriptManager
import world.guild.Guild
import world.ranking.RankingKeeper
import java.util.*
import java.util.concurrent.TimeUnit
import java.util.function.Function

abstract class AbstractPlayerInteraction(val c: Client) {

    val player: Character get() = c.character

    fun gainItem(id: Int, quantity: Int) {
        gainItemInternal(id, quantity)
        player.localEffect(QuestEffect(id, quantity))
    }

    fun massGainItem(items: Array<IntArray>) {
        val entries: MutableList<Pair<Int, Int>> = ArrayList()
        Arrays.stream(items).forEach { item: IntArray ->
            gainItemInternal(item[0], item[1])
            entries.add(Pair.create(item[0], item[1]))
        }
        player.localEffect(QuestEffect(entries))
    }

    private fun gainItemInternal(id: Int, quantity: Int) {
        if (quantity > 0) {
            val item = ItemManager.getItem(id)
            player.modifyInventory({ it.add(item, quantity.toShort()) })
        } else {
            player.modifyInventory({ it.take(id, (-quantity).toShort()) })
        }
    }

    fun gainMeso(meso: Int) {
        player.gainMeso(meso, true)
    }

    fun haveItem(id: Int, quantity: Int): Boolean {
        return player.getItemQuantity(id) >= quantity
    }

    fun gainExp(exp: Int) {
        player.gainExp(exp)
        val msg = IncEXPMessage()
        msg.exp = exp
        msg.onQuest = true
        player.message(msg)
    }

    fun warp(id: Int, portal: String) {
        player.changeField(id, portal)
    }

    fun warp(id: Int) {
        player.changeField(id)
    }

    val mapId: Int get() = player.fieldId

    fun sendBlue(message: String) {
        player.message(NoticeWithoutPrefixMessage(message))
    }

    fun sendNpcMessage(message: String, npc: Int) {
        player.message(UtilDlgExMessage(message, npc))
    }

    fun setQuestProgress(qid: Int, mob: Int, progress: String) {
        val quest: Quest = player.quests[qid] ?: return
        quest.progress(mob, progress.toInt())
    }

    fun getQuest(id: Int): Quest? {
        return player.quests[id]
    }

    fun openNpc(id: Int) {
        if (!NPCScriptManager.converse(c, id)) {
            sendBlue("Npc does not have a script")
        }
    }

    fun isEquip(id: Int): Boolean {
        val template = ItemManager.getItem(id)
        return template is ItemEquipTemplate
    }

    fun getEquipById(id: Int): ItemSlotEquip? {
        val template = ItemManager.getItem(id)
        return template.toItemSlot() as ItemSlotEquip?
    }

    fun gainStatItem(id: Int, obj: Any) {
        val stats: AbstractMap<String, Int> = obj as AbstractMap<String, Int>
        val equip = getEquipById(id)
        val template = ItemManager.getItem(id) as ItemEquipTemplate
        if (equip != null) {
            equip.str = stats["STR"]?.toShort() ?: template.incSTR
            equip.dex = stats["DEX"]?.toShort() ?: template.incDEX
            equip.luk = stats["LUK"]?.toShort() ?: template.incLUK
            equip.int = stats["INT"]?.toShort() ?: template.incINT
            equip.pad = stats["PAD"]?.toShort() ?: template.incPAD
            equip.mad = stats["MAD"]?.toShort() ?: template.incMAD
            equip.acc = stats["ACC"]?.toShort() ?: template.incACC
            equip.eva = stats["EVA"]?.toShort() ?: template.incEVA
            equip.jump = stats["JUMP"]?.toShort() ?: template.incJump
            equip.speed = stats["SPEED"]?.toShort() ?: template.incSpeed
            equip.pdd = stats["PDD"]?.toShort() ?: template.incPDD
            equip.mdd = stats["MDD"]?.toShort() ?: template.incMDD
            equip.maxHP = stats["HP"]?.toShort() ?: template.incMaxHP.toShort()
            equip.maxMP = stats["MP"]?.toShort() ?: template.incMaxMP.toShort()
            equip.ruc = stats["SLOTS"]?.toByte() ?: 7
            player.modifyInventory({ it.add(equip) })
        }
    }

    fun tremble(heavy: Boolean, delay: Int) {
        player.field.fieldEffect(TrembleFieldEffect(heavy, delay))
    }

    fun openNpcIn(npc: Int, time: Int, dispose: Boolean = true) {
        c.ch.eventLoop().schedule({
            if (dispose) {
                c.lastNpcClick = 0
                NPCScriptManager.dispose(c)
            }
            NPCScriptManager.converse(c, npc)
        }, time.toLong(), TimeUnit.MILLISECONDS)
    }

    fun executeAfter(func: Function<AbstractPlayerInteraction?, Void?>, after: Int) {
        c.ch.eventLoop().schedule<Void?>({ func.apply(this) }, after.toLong(), TimeUnit.MILLISECONDS)
    }

    fun alert(msg: String) {
        player.message(AlertMessage(msg))
    }

    fun openStorage(npcId: Int) {
        ItemStorageInteraction(npcId, c.storage).open(player)
    }

    fun getGuild(): Guild? {
        return player.guild
    }

    fun changeGuildName(name: String) {
        GuildPackets.changeGuildName(player, name)
    }

    fun loadGuild() {
        player.guild?.let {
            c.write(it.getLoadGuildPacket())
        }
    }

    fun getMobTemplate(id: Int): FieldMobTemplate {
        return MobManager.getMob(id)
    }

    fun getMap(): Field {
        return player.field
    }

    val mobsOnField: List<Any>
        get() {
            val mobs: MutableList<FieldMob> = ArrayList<FieldMob>()
            c.character.field.getObjects<FieldMob>().forEach { mobs.add(it) }
            return mobs
        }

    fun getRankings(): RankingKeeper {
        return RankingKeeper
    }

    fun getServer(): Server {
        return Server
    }

    fun getCharacters(): List<Character> {
        return player.field.getObjects<Character>().toList()
    }

    fun getMobs(): List<FieldMob> {
        return player.field.getObjects<FieldMob>().toList()
    }

    fun getNpcs(): List<FieldNPC> {
        return player.field.getObjects<FieldNPC>().toList()
    }

    fun getDrops(): List<AbstractFieldDrop> {
        return player.field.getObjects<AbstractFieldDrop>().toList()
    }

    fun getMobByObjId(id: Int): FieldMob? {
        return player.field.getObject(id)
    }
}